Adds shadow piercing combinators

leonsenft opened the pull request:
Adds support for shadow piercing combinator `/deep/` and its alias `>>>`.

Fixes dart-lang/tools#765

leonsenft committed:
Adds support for shadow piercing combinators

Fixes #23

terrylucas commented on lib/parser.dart:
```diff
@@ -1229,13 +1229,30 @@ class _Parser {
+        _eat(TokenKind.SLASH);
+        var ate = _maybeEat(TokenKind.IDENTIFIER);
+        var tok = ate ? _previousToken : _peekToken;
+        if (!(ate && tok.text == 'deep')) {
```
Do you want to support /deep/ ?  I noticed its no longer mentioned in the spec https://drafts.csswg.org/css-scoping-1/#deep-combinator only >>> is the deep combinator.

If you need to support 'deep' then all keywords/reserved words e.g., 'deep' should be in the TokenKind not hard coded in the parser.  Once CSS exposes a name its usually a good chance they'll add more reserved names.  You might do something like:

static const List<Map<String, dynamic>> COMBINATOR_NAMES = const [
    const {'type': TokenKind.COMBINATOR_DEEP, 'value': 'deep'},
];

  static int matchCombinatorNames(String text, int offset, int length) {
    return matchList(COMBINATOR_NAMES, 'type', text, offset, length);
  }

then the parser would have this code:

case TokenKind.SLASH:
     // Parse /deep/
     _eat(TokenKind.SLASH);
     if (_peekIdentifier()) {
        // Is it a directive?
        var combinatorName = _peekToken.text;
        var combinatorLen = directive.length;
        tokId = TokenKind.matchCombinatorNames(combinatorName, 0, combinatorLen);
...

leonsenft replied:
We should support `/deep/` for backwards compatibility as up until recently `>>>` was merely an alias. We could deprecate `/deep/` and issue a warning instructing users to switch to `>>>` if desired.

I'm skeptical we'll see any combinators of this form with reserved names. The purposed [reference combinator](https://www.w3.org/TR/selectors4/#idref-combinators) used this syntax, but has since been [removed](https://drafts.csswg.org/selectors/#changes).

zoechi replied:
SASS supports `/deep/` but not `>>>`, therefore I'd prefer it would just be kept. `>>>` is as deprecated as `/deep/` AFAIK. Perhaps it can be used in `querySelector()` but not in styles.

leonsenft replied:
@zoechi @terrylucas Maybe the best solution would be to create separate token kinds for each combinator, that way the original styles are preserved so we don't break compatibility with any other tools?

terrylucas commented on lib/src/tree.dart:
```diff
@@ -120,12 +120,24 @@ class SimpleSelectorSequence extends TreeNode {
+  String get _combinatorToString {
+    switch (combinator) {
+      case TokenKind.COMBINATOR_DEEP:
+        return ' /deep/ ';
```
Should >>> be the default since that's the only deep operator for combinators in the latest spec?

leonsenft replied:
Will do!

terrylucas commented on lib/src/tree_printer.dart:
```diff
@@ -271,6 +271,8 @@ class _TreePrinter extends Visitor {
     } else if (node.isCombinatorTilde) {
       output.writeValue('combinator', "~");
+    } else if (node.isCombinatorDeep) {
+      output.writeValue('combinator', '/deep/');
```
Ditto, should >>> be the default?

terrylucas commented:
Just a couple of questions?

leonsenft committed:
Swaps >>> and /deep/ as deep combinator and alias

The latest spec at this time has removed /deep/ altogether and promoted >>> from
an alias for /deep/: https://drafts.csswg.org/css-scoping-1/#deep-combinator

/deep/ will remain as an alias for backwards compatibility.

terrylucas reviewed (approved):
lgtm

terrylucas commented:
Unfortunate, that your stuck with preliminary-preliminary syntax that is
superseded by newer preliminary syntax.  Maybe a TODO should be added to
jettison /deep/ for the new >>> goodness.

On Wed, Jan 25, 2017 at 12:05 PM, Leon Senft <notifications@github.com>
wrote:

> *@leonsenft* commented on this pull request.
> ------------------------------
>
> In lib/parser.dart <https://github.com/dart-lang/csslib/pull/30>:
>
> >          break;
>        case TokenKind.TILDE:
>          _eat(TokenKind.TILDE);
>          combinatorType = TokenKind.COMBINATOR_TILDE;
>          break;
> +      case TokenKind.SLASH:
> +        // Parse /deep/
> +        _eat(TokenKind.SLASH);
> +        var ate = _maybeEat(TokenKind.IDENTIFIER);
> +        var tok = ate ? _previousToken : _peekToken;
> +        if (!(ate && tok.text == 'deep')) {
>
> @zoechi <https://github.com/zoechi> @terrylucas
> <https://github.com/terrylucas> Maybe the best solution would be to
> create separate token kinds for each combinator, that way the original
> styles are preserved so we don't break compatibility with any other tools?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dart-lang/csslib/pull/30>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB9ewWv4QhqzafybgmyuoQdRmg0J3CcDks5rV6sNgaJpZM4LptNx>
> .
>

leonsenft committed:
'>>>' and '/deep/' no longer share a token

This allows csslib to preserve the original combinator choice when printing an
AST.
